### PR TITLE
amp-list: 'set-dom' for DOM diffing

### PR DESCRIPTION
**Pull request by choumx**
Size impact: 7001 - 5590 = 1411 bytes (gzip)

May help with "white flash" experienced when re-rendering via `[src]` binding, e.g. https://github.com/ampproject/amphtml/issues/14798#issuecomment-395273608

### CODE REVIEW

File: extensions/amp-list/0.1/amp-list.js [hunk 1]
```diff
@@ -375,13 +398,21 @@ export class AmpList extends AMP.BaseElement {
...
-        }
-        this.container_.appendChild(element);
-      });
+      const diffing = isExperimentOn(this.win, 'amp-list-diffing');
+      if (this.container_.hasChildNodes() && diffing) {
+        const newContainer = this.createContainer_();
+        this.addElementsToContainer_(elements, newContainer);
+
+        // Necessary to support both browserify and CC import semantics.
+        const diff = (setDOM.default || setDOM);
```
Comment (choumx) [inline extensions/amp-list/0.1/amp-list.js]: This is pretty ugly. Any ideas?

**Comment by choumx** [comment]
Friendly ping.

File: extensions/amp-list/0.1/amp-list.js [hunk 2]
```diff
@@ -375,13 +398,21 @@ export class AmpList extends AMP.BaseElement {
...
+        // Use `i-amphtml-key` as a node key for identifying when to skip
+        // DOM diffing and replace. Needed for AMP elements, for example.
+        diff.KEY = 'i-amphtml-key';
+        diff(this.container_, newContainer);
```
Comment (jridgewell) [inline extensions/amp-list/0.1/amp-list.js]: Need tests. How does this work when encountering `<amp-img`, etc.

File: src/purifier.js [hunk 3]
```diff
@@ -248,6 +255,14 @@ export function purifyHtml(dirty) {
   let allowedAttributes;
   const allowedAttributesChanges = [];
 
+  // Disables DOM diffing for a given node and allows it to be replaced.
+  const disableDiffingFor = node => {
+    if (diffing) {
+      // set-dom uses node attribute keys for opting out of diffing.
+      node.setAttribute('i-amphtml-key', KEY_COUNTER++);
```
Comment (jridgewell) [inline src/purifier.js]: Why isn't this using `setDOM.IGNORE`?

File: src/purifier.js [hunk 4]
```diff
@@ -351,6 +368,8 @@ export function purifyHtml(dirty) {
       // Set a custom attribute to mark this element as containing a binding.
       // This is an optimization that obviates the need for DOM scan later.
       node.setAttribute('i-amphtml-binding', '');
+      // Avoid disrupting Bind.scanAndApply() with DOM diffing.
+      disableDiffingFor(node);
```
Comment (jridgewell) [inline src/purifier.js]: Can you explain why?

File: src/sanitizer.js [hunk 5]
```diff
@@ -175,7 +175,19 @@ function sanitizeWithCaja(html) {
...
       }
-      let emittedBindingMarker = false;
+      // Filter out bindings with empty attribute values.
+      const hasBindings = bindingAttribs.some(i => !!attribs[i + 1]);
+      if (hasBindings) {
+        // Set a custom attribute to identify elements with bindings.
+        // This is an optimization that avoids the need for a DOM scan later.
+        attribs.push('i-amphtml-binding', '');
+      }
+      // Elements with bindings and AMP elements must opt-out of DOM diffing.
```
Comment (jridgewell) [inline src/sanitizer.js]: Can you explain why?

File: extensions/amp-list/0.1/amp-list.js [hunk 6]
```diff
@@ -406,17 +435,27 @@ export class AmpList extends AMP.BaseElement {
...
-      removeChildren(dev().assertElement(this.container_));
-      elements.forEach(element => {
-        if (!element.hasAttribute('role')) {
-          element.setAttribute('role', 'listitem');
-        }
-        this.container_.appendChild(element);
-      });
+
+      const diffing = isExperimentOn(this.win, 'amp-list-diffing');
+      if (this.container_.hasChildNodes() && diffing) {
```
Comment (jridgewell) [inline extensions/amp-list/0.1/amp-list.js]: Nit: swap the order

File: src/purifier.js [hunk 7]
```diff
@@ -355,6 +372,10 @@ export function purifyHtml(dirty) {
       // Set a custom attribute to mark this element as containing a binding.
       // This is an optimization that obviates the need for DOM scan later.
       node.setAttribute('i-amphtml-binding', '');
+      // Don't DOM diff nodes with bindings because amp-bind scans newly
+      // rendered elements and discards _all_ old elements _before_ diffing, so
+      // preserving some old elements would cause loss of functionality.
+      disableDiffingFor(node);
```
Comment (jridgewell) [inline src/purifier.js]: This can be called multiple times on a node, but it's unnecessary after the first time.

**Comment by choumx** [comment]
@cathyxz I think we'll launch this under an attribute opt-in but we should probably enable by default in the redesigned `<amp-render>`.

**Pull request merged**